File: python/algorithms/dynamic_programming/digit/digit_dp.py
"""数位 DP：统计 [0, n] 中数位和 % K == 0 的整数个数（修正前导零）。"""

from __future__ import annotations
from functools import lru_cache


def count_digit_sum_mod0(n: int, k: int) -> int:
    if k <= 0:
        raise ValueError("k must be positive")
    if n < 0:
        return 0
    s = str(n)

    @lru_cache(maxsize=None)
    def dfs(i: int, tight: bool, mod: int, z: bool) -> int:
        """z=True 表示仍处于前导零阶段（数值尚未开始）。"""
        if i == len(s):
            return 1 if mod == 0 else 0
        limit = int(s[i]) if tight else 9
        total = 0
        for d in range(0, limit + 1):
            ntight = tight and (d == limit)
            nz = z and (d == 0)
            nmod = mod if nz else (mod + d) % k
            total += dfs(i + 1, ntight, nmod, nz)
        return total

    return dfs(0, True, 0, True)


if __name__ == "__main__":
    k = 3

    def brute(nn: int) -> int:
        c = 0
        for x in range(0, nn + 1):
            if sum(int(ch) for ch in str(x)) % k == 0:
                c += 1
        return c

    for nn in range(0, 500):
        assert count_digit_sum_mod0(nn, k) == brute(nn)
    try:
        count_digit_sum_mod0(10, 0)
        raise AssertionError("expected ValueError")
    except ValueError:
        pass
    print("digit_dp OK")

File: cpp/algorithms/dynamic_programming/digit/digit_dp.cpp
#include <alg_std.hpp>
#include <cassert>
using namespace std;

int K;
string s;
int memo[25][2][10][2];

int dfs(int i, bool tight, int mod, int z) {
    if (i == (int)s.size()) return mod == 0 ? 1 : 0;
    int& mem = memo[i][tight][mod][z];
    if (mem != -1) return mem;
    int lim = tight ? s[i] - '0' : 9;
    int tot = 0;
    for (int d = 0; d <= lim; ++d) {
        bool ntight = tight && (d == lim);
        bool nz = z && (d == 0);
        int nmod = nz ? mod : (mod + d) % K;
        tot += dfs(i + 1, ntight, nmod, nz);
    }
    return mem = tot;
}

int count_digit_sum_mod0(long long n, int k) {
    if (k <= 0) throw runtime_error("k must be positive");
    if (n < 0) return 0;
    K = k;
    s = to_string(n);
    memset(memo, -1, sizeof(memo));
    return dfs(0, true, 0, true);
}

int brute(long long nn, int k) {
    int c = 0;
    for (long long x = 0; x <= nn; ++x) {
        int sm = 0;
        for (char ch : to_string(x)) sm += ch - '0';
        if (sm % k == 0) ++c;
    }
    return c;
}

int main() {
    int k = 3;
    for (long long nn = 0; nn < 500; ++nn) assert(count_digit_sum_mod0(nn, k) == brute(nn, k));
    try {
        count_digit_sum_mod0(10, 0);
        assert(false);
    } catch (const runtime_error&) {
    }
    cout << "digit_dp OK" << endl;
    return 0;
}
